src/hpr_routing.erl

```diff
@@ -77,87 +91,127 @@ find_routes({join_req, {AppEUI, DevEUI}}, _Packet) ->
+    }
 ) -> {ok, [{hpr_route:route(), non_neg_integer()}]} | {error, invalid_mic}.
-find_routes_for_uplink(#frfu{routes = [], no_skf_routes = []}, '$end_of_table') ->
+find_routes_for_uplink(_Packet, _DevAddr, [], [], undefined) ->
     {error, invalid_mic};
-find_routes_for_uplink(#frfu{routes = Routes, no_skf_routes = NoSKFRoutes}, '$end_of_table') ->
-    {ok, lists:usort(Routes ++ [{R, 0} || R <- NoSKFRoutes])};
+find_routes_for_uplink(_Packet, _DevAddr, [], EmptyRoutes, undefined) ->
+    {ok, [{R, 0} || R <- EmptyRoutes]};
 find_routes_for_uplink(
-    #frfu{
-        payload = Payload,
-        routes = Routes,
-        dev_range_routes = DevRangeRoutes,
-        no_skf_routes = NoSKFRoutes0
-    } = FRFU,
-    {SKFs, Continuation}
+    _Packet, DevAddr, [], EmptyRoutes, {Route, SessionKey, LastUsed, SKFMaxCopies}
 ) ->
-    MicFilteredRoutes = lists:filtermap(
-        fun({Key, RouteID, SKFMaxCopies}) ->
-            case hpr_lorawan:key_matches_mic(Key, Payload) of
-                %% If failed we remove that entry
-                false ->
-                    false;
-                %% If match we try to get the Route and included it.
+    LastHour = erlang:system_time(millisecond) - ?SKF_UPDATE,
+    case LastUsed < LastHour of
+        true ->
+            erlang:spawn(hpr_route_ets, update_skf, [
+                DevAddr, SessionKey, hpr_route:id(Route), SKFMaxCopies
+            ]);
+        false ->
+            ok
+    end,
+    {ok, [{Route, SKFMaxCopies} | [{R, 0} || R <- EmptyRoutes]]};
+find_routes_for_uplink(Packet, DevAddr, [{Route, SKFETS} | Routes], EmptyRoutes, SelectedRoute) ->
+    case check_route_skfs(Packet, DevAddr, SKFETS) of
+        empty ->
+            case hpr_route:ignore_empty_skf(Route) of
                 true ->
-                    case hpr_route_ets:lookup_route(RouteID) of
-                        %% Note: this would mark packet as invalid_mic
-                        %% but we are just missing the route
-                        [] -> false;
-                        Rs -> {true, [{R, SKFMaxCopies} || R <- Rs]}
-                    end
+                    find_routes_for_uplink(Packet, DevAddr, Routes, EmptyRoutes, SelectedRoute);
+                false ->
+                    find_routes_for_uplink(
+                        Packet, DevAddr, Routes, [Route | EmptyRoutes], SelectedRoute
+                    )
+            end;
+        false ->
+            find_routes_for_uplink(Packet, DevAddr, Routes, EmptyRoutes, SelectedRoute);
+        {ok, SessionKey, LastUsed, SKFMaxCopies} ->
+            case SelectedRoute of
+                undefined ->
+                    find_routes_for_uplink(
+                        Packet,
+                        DevAddr,
+                        Routes,
+                        EmptyRoutes,
+                        {Route, SessionKey, LastUsed, SKFMaxCopies}
+                    );
+                {_R, _S, L, _M} when LastUsed > L ->
```

Comment:
This is...
We had a `SelectedRoute`, and found another applicable route. 
`SelectedRoute` get's swapped out if the new route has been used more recently.

We're not returning early when we find a first match, or delivering to multiple routes if for some reason they put the same (devaddr, session_key) in more than one place.

If the same (devaddr, session_key) is in more than one route, we could end up swapping where we send the packet every hour, which would be very confusing on many fronts.

Reply:
> We're not returning early when we find a first match, or delivering to multiple routes if for some reason they put the same (devaddr, session_key) in more than one place.

No because config service still allows same session keys in different routes so until we can address that we will have to check every route, I could pass the timestamp found to minimize select for other routes.

> If a (devaddr, session_key) pair was added to a different route, it would hijack the existing routing. And after a restart, routing would be dependent on db ordering from the config service.

Yes true, single session key will fix that.